docs/decisions/0007-enforcement-mechanisms-mfe.rst

```diff
@@ -0,0 +1,200 @@
+0007: Enforcement mechanisms - current user permission checks from MFEs and other remote clients
+################################################################################################
+
+Status
+********
+
+**Draft**
+
+Context
+*********
+
+Authorization (AuthZ) decisions need to be enforced not only on the
+backend services, but also on the frontend applications, including
+Micro-Frontends (MFEs) and mobile apps. This ensures that users only see
+and can interact with the UI elements they are permitted to access based
+on their roles and permissions.
+
+To achieve this, we need to establish effective communication mechanisms
+between the backend services and the frontend clients. This involves
+determining how AuthZ data is transmitted, how often it is updated, and
+how the frontend applications can use this data to enforce access
+control for the currently authenticated user.
+
+We want an approach that:
+
+-  Minimizes the amount of requests and data transferred between the
+   backend and frontend.
+-  Is granular enough to match the functionality available on the
+   backend.
+-  Is easy to implement and maintain across different frontend
+   applications.
+
+**Please note:** The scope of this ADR is limited to enforcing permissions
+for the currently authenticated user.
+
+Decision
+**********
+
+I. REST API for authorization queries
+=====================================
+
+We will implement a dedicated REST API endpoint that frontend
+applications can use to query for specific permissions for the currently
+authenticated user. Queries would be at the Subject-Action-Object-Context
+level, being the Subject always the current authenticated user. This allows
+the frontend to check if a user has permission to perform a specific action
+on a specific object within a given context.
+
+To optimize performance and reduce latency, the API will support batch
+queries, allowing multiple permission checks in a single request. It
+will also have caching mechanisms in place.
+
+API Definition
+--------------
+
+POST /api/authz/v1/permissions/validate/me
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Validate if the current user has specific permissions.
+
+Request Body:
+"""""""""""""
+
+Format:
+
+.. code:: ts
+
+   Array<{
+       action: string
+       object?: string
+       scope?: string
+   }>
+
+Example:
+
+.. code:: json
+
+   [
+       {
+           "action": "act:read",
+           "object": "lib:test-lib",
+           "scope": "org:OpenedX"
+       },
+       {
+           "action": "act:edit",
+           "object": "lib:test-lib",
+           "scope": "org:OpenedX"
+       }
+   ]
```

Comment:
Does this mean that if a user has permission on a scope, they automatically have permission on the object within it? If so, our decision is that scope inheritance will not be managed implicitly. Instead, checks must explicitly handle containment.

For example:

```
can(user, act:read, lib:test-lib) OR can(user, act:read, org:OpenedX)
```

Here, `lib:test-lib` and `org:OpenedX` are treated as separate scopes. I’m not sure if this belongs directly in this ADR, but it’s worth noting for clarity. Still I think this ADR covers this case as well, since object is optional.

FYI @bmtcril

Reply:
@rodmgwgu: that's the level of granularity we're going with for now. It's not set in stone, since the model could later evolve to be more granular. For now, let's note that possibility but focus on describing the current, more feasible approach.